GVFS/GVFS.Virtualization/FileSystemCallbacks.cs

```diff
@@ -506,6 +518,34 @@ private static bool CheckConditionWithRetry(Func<bool> predicate, int retries, i
             return result;
         }
 
+        private EventMetadata GetProcessInteractionData(ConcurrentDictionary<string, PlaceHolderCreateCounter> collectedData, ref bool logToFile)
+        {
+            EventMetadata metadata = new EventMetadata();
+
+            if (collectedData.Count > 0)
+            {
+                int count = 0;
+                foreach (KeyValuePair<string, PlaceHolderCreateCounter> processCount in
+                collectedData.OrderByDescending((KeyValuePair<string, PlaceHolderCreateCounter> kvp) => kvp.Value.Count).Take(10))
+                {
+                    ++count;
+                    metadata.Add("ProcessName" + count, processCount.Key);
+                    metadata.Add("ProcessCount" + count, processCount.Value.Count);
+                }
+
+                logToFile = true;
+            }
+
+            return metadata;
+        }
+
+        private ConcurrentDictionary<string, PlaceHolderCreateCounter> GetAndResetProcessCountMetadata(ref ConcurrentDictionary<string, PlaceHolderCreateCounter> collectedData)
+        {
+            ConcurrentDictionary<string, PlaceHolderCreateCounter> localData = collectedData;

```

Comment:
I understand *why* we are doing this, but I think it is confusing.

Maybe a helpful comment at the top like:
// Captures the current state of dictionary, and resets it
// This approach is optimal for our use case to preserve all entries while avoiding additional locking

Because what I think you want in a perfect world is to lock the 3 dictionaries why you examine the data.  But I can see that being a pain and more code verbosity/complexity.